### ISSUE TITLE
Performance slowdown and high memory consumption when using dynamic imports

### ISSUE THREAD

**Issue by meskill** [webpack/webpack#17416]
# Bug report

When using dynamic `import` call over sync `require` I see increase in build time and in memory consumption.

**What is the current behavior?**

I have some private application that uses dynamic imports to implement lazy loading.

After some code simplification and enabling additional logs, If I run it I see next numbers for time and memory:
```
DEBUG LOG from webpack.buildChunkGraph
<t> visitModules: prepare: 2.5296 ms
<t> visitModules: visiting: 6.703999 ms
<t> visitModules: calculating available modules: 0.6491 ms
<t> visitModules: merging available modules: 0.939 ms
<t> visitModules: check modules for revisit: 0.3369 ms
<t> visitModules: prepare: 45.050496 ms
<t> visitModules: visiting: 778.289864 ms
    549028 queue items processed (265085 blocks)
    380 chunk groups connected
    380 chunk groups processed for merging (380 module sets, 0 forked, 0 + 0 modules forked, 0 + 0 modules merged into fork, 0 resulting modules)
    380 chunk group info updated (0 already connected chunk groups reconnected)
<t> visitModules: 788.805262 ms
<t> connectChunkGroups: 1.801699 ms
<t> cleanup: 0.1743 ms

client (webpack 5.87.0) compiled with 1 warning in 110049 ms
{
  clientBuildTime: 110050,
  maxMemoryRss: 2221084672
}
```

If I replace dynamic imports with sync requires I get this:
```
DEBUG LOG from webpack.buildChunkGraph
<t> visitModules: prepare: 113.764402 ms
<t> visitModules: visiting: 187.628703 ms
    4353 queue items processed (2031 blocks)
    0 chunk groups connected
    0 chunk groups processed for merging (0 module sets, 0 forked, 0 + 0 modules forked, 0 + 0 modules merged into fork, 0 resulting modules)
    0 chunk group info updated (0 already connected chunk groups reconnected)
<t> visitModules: 189.600003 ms
<t> connectChunkGroups: 0.2773 ms
<t> cleanup: 0.0831 ms

client (webpack 5.87.0) compiled with 3 warnings in 100983 ms
{
  clientBuildTime: 100984,
  maxMemoryRss: 1147891712
}
```

You can see that build with dynamic imports executing for 10 seconds longer and requires almost twice of memory. 

**If the current behavior is a bug, please provide the steps to reproduce.**

I've created an [reproduction repo](https://github.com/meskill/webpack5-performance-dynamic-import) that shows the issue, although the numbers are not that high as in real application.

**What is the expected behavior?**

I hope there could be a room to improve this to make lazy chunking less expensive for building. 

For example in case of dynamic imports the more time is spend in `visitModules` function that has [comment](https://github.com/webpack/webpack/blob/8241da7f1e75c5581ba535d127fa66aeb9eb2ac8/lib/buildChunkGraph.js#L541) about possibility of removing some traversals.

**Other relevant information:**
webpack version:  5.87.0
Node.js version: 16
Operating System: Debian bullseye

**Comment by jdfm** [comment]
tl;dr: This seems to be a regression in NodeJS that was introduced in NodeJS 16.11 after they upgraded the V8 engine, it's still an issue.

Details:

We've been running into memory issues in our NextJS build (latest v12) and some of our Jest test suites. The common factor between the build process and the test suites are the presence of dynamic imports in the code. The test suites that don't have dynamic imports, don't seem to have memory issues, the ones that do, do. 

We eventually noticed the following Jest issue that's currently still open https://github.com/jestjs/jest/issues/11956 and the fix mentioned there (`--workerIdleMemoryLimit` jest argument) seemed to improve things for us on the Jest side of things. As for NextJS we're currently attempting to resolve this by tweaking the `--max-old-space-size` we pass to node via `NODE_OPTIONS` and making sure we try to calculate it based on the maximum amount of memory and cpu resources we're making available during the build process, as the NODE_OPTIONS in our case get distributed to a number of NodeJS processes that NextJS seems to spawn (could also be Webpack doing that under the hood, we didn't dig too too deeply into that side of things).

Would it be possible for Webpack to implement a similar option until the issue in NodeJS gets resolved for good?

**Comment by alexander-akait** [comment]
Looks like it was fixed, just tested on Node.js@16.15.0 and don't see the problem anymore, so I want to close and sorry for delay, anyway feel free to feedback
